File: gregorio/lib/movies_google/movies/movie.ex
defmodule MoviesGoogle.Movies.Movie do
  use Ecto.Schema
  import Ecto.Changeset

  schema "movies" do
    field :tmdb_id, :integer
    field :title, :string
    field :original_title, :string
    field :overview, :string
    field :tagline, :string
    field :genres, :string
    field :release_date, :date
    field :runtime, :integer
    field :vote_average, :float
    field :vote_count, :integer
    field :popularity, :float
    field :budget, :integer
    field :revenue, :integer
    field :original_language, :string
    field :director, :string
    field :cast_names, :string
    field :embedding, :binary

    timestamps()
  end

  def changeset(movie, attrs) do
    movie
    |> cast(attrs, [
      :tmdb_id, :title, :original_title, :overview, :tagline,
      :genres, :release_date, :runtime, :vote_average, :vote_count,
      :popularity, :budget, :revenue, :original_language,
      :director, :cast_names
    ])
    |> validate_required([:tmdb_id, :title])
    |> unique_constraint(:tmdb_id)
  end
end

File: fabiano/backend/src/models/Movie.js
import mongoose from 'mongoose';

const genreSchema = new mongoose.Schema(
  {
    id: Number,
    name: String,
  },
  { _id: false }
);

const castPreviewSchema = new mongoose.Schema(
  {
    name: String,
    character: String,
  },
  { _id: false }
);

const movieSchema = new mongoose.Schema(
  {
    tmdbId: { type: Number, required: true, unique: true, index: true },
    title: { type: String, required: true },
    originalTitle: String,
    overview: String,
    tagline: String,
    releaseDate: Date,
    runtime: Number,
    voteAverage: Number,
    voteCount: Number,
    popularity: Number,
    genres: [genreSchema],
    genreNames: String,
    keywordNames: String,
    castPreview: [castPreviewSchema],
    castNames: String,
    directorNames: String,
    homepage: String,
    status: String,
    budget: Number,
    revenue: Number,
    originalLanguage: String,
  },
  { timestamps: true }
);

movieSchema.index(
  {
    title: 'text',
    originalTitle: 'text',
    tagline: 'text',
    overview: 'text',
    genreNames: 'text',
    keywordNames: 'text',
    castNames: 'text',
    directorNames: 'text',
  },
  {
    weights: {
      title: 10,
      originalTitle: 8,
      tagline: 6,
      genreNames: 4,
      keywordNames: 3,
      castNames: 5,
      directorNames: 5,
      overview: 1,
    },
    name: 'movie_text_index',
  }
);

movieSchema.index({ title: 1 });
movieSchema.index({ releaseDate: -1 });
movieSchema.index({ popularity: -1 });
movieSchema.index({ voteAverage: -1 });

export const Movie = mongoose.model('Movie', movieSchema);
